warehouse/ingest-core/src/main/java/datawave/ingest/util/cache/lease/JobCacheZkLockFactory.java

```diff
@@ -0,0 +1,277 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/** Factory class to handle job cache leases. */
+public class JobCacheZkLockFactory implements JobCacheLockFactory {
+    private static final Logger LOGGER = LoggerFactory.getLogger(JobCacheZkLockFactory.class);
+    
+    public static final String DW_JOB_CACHE_ZOOKEEPER_TIMEOUT = "datawave.job.cache.zk.timeout.ms";
+    public static final String DW_JOB_CACHE_ZOOKEEPER_RETRY_CNT = "datawave.job.cache.zk.retry.cnt";
+    public static final String DW_JOB_CACHE_ZOOKEEPER_RETRY_TIMEOUT = "datawave.job.cache.zk.retry.timeout.ms";
+    public static final String DW_JOB_CACHE_ZOOKEEPER_NAMESPACE = "datawave.job.cache.zk.namespace";
+    public static final String DW_JOB_CACHE_ZOOKEEPER_NODES = "datawave.job.cache.zk.nodes";
+    private static final int ZOOKEEPER_TIME_BEFORE_RETRY_MS = 30;
+    private static final int ZOOKEEPER_ACQUIRE_LOCK_TIMEOUT = 30;
+    private static final int ZOOKEEPER_RETRY_CONNECTION_CNT = 3;
+    
+    private static final String ZOOKEEPER_LEASE_DIR = "/leases";
+    
+    private CuratorFramework client;
+    private final Map<String,InterProcessSemaphoreMutex> jobIdToLeases = Maps.newHashMap();
+    private int lockTimeoutInMs;
+    
+    public JobCacheZkLockFactory() {}
+    
+    public JobCacheZkLockFactory(String namespace, String zookeepers, int lockTimeoutInMs, int retryCnt, int retryTimeoutInMS) {
+        init(namespace, zookeepers, lockTimeoutInMs, retryCnt, retryTimeoutInMS);
+    }
+    
+    @Override
+    public void init(Configuration conf) {
+        int lockTimeout = conf.getInt(DW_JOB_CACHE_ZOOKEEPER_TIMEOUT, ZOOKEEPER_TIME_BEFORE_RETRY_MS);
+        int retryCnt = conf.getInt(DW_JOB_CACHE_ZOOKEEPER_RETRY_CNT, ZOOKEEPER_RETRY_CONNECTION_CNT);
+        int retryTimeout = conf.getInt(DW_JOB_CACHE_ZOOKEEPER_RETRY_TIMEOUT, ZOOKEEPER_ACQUIRE_LOCK_TIMEOUT);
+        String namespace = conf.get(DW_JOB_CACHE_ZOOKEEPER_NAMESPACE);
+        String zookeepers = conf.get(DW_JOB_CACHE_ZOOKEEPER_NODES);
+        init(namespace, zookeepers, lockTimeout, retryCnt, retryTimeout);
+    }
+    
+    @Override
+    public boolean acquireLock(String id) {
+        return acquireLock(id, this::isCacheLocked);
+    }
+    
+    @Override
+    public boolean acquireLock(String id, Predicate<String> cacheCheck) {
+        if (cacheCheck.test(getCacheId(id))) {
+            LOGGER.warn("Unable to acquire job lease for {} since cache is locked or active if deleting", id);
+            return false;
+        }
+        
+        InterProcessSemaphoreMutex lock = new InterProcessSemaphoreMutex(client, getLockPath(id));
+        try {
+            if (lock.acquire(lockTimeoutInMs, TimeUnit.MILLISECONDS)) {
+                jobIdToLeases.put(id, lock);
+                return true;
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Unable to acquire lock for {} ", id, e);
```

Comment:
This actually maps to `Logger.warn(String message, Object... arguments)`, so you won't get the nice stacktrace output. Instead, I think we want `LOGGER.warn("Unable to acquire lock for " + id, e)` to get the `Logger.warn(String message, Throwable cause)` functionality.

Reply:
Thanks!  I may have made this mistake in multiple places.  I will take a look back as well.

Reply:
Actually, I think I'm wrong here. It looks like Log4j might actually examine the last item to see if it extends Throwable. My bad!